R/model-tools.R

```diff
@@ -371,3 +371,92 @@ enw_model <- function(model = system.file(
   }
   return(model)
 }
+
+#' Expose `epinowcast` stan functions in R
+#'
+#' @description This function facilitates the exposure of Stan functions from
+#' the [epinowcast]() package in R. It utilizes the [expose_functions()] method
+#' of [cmdstanr::CmdStanModel] or this purpose. This function is useful for
+#' developers and contributors to the [epinowcast] package, as well as for
+#' users interested in exploring and prototyping with model functionalities.
+#'
+#' @param files A character vector specifying the names of Stan files to be
+#' exposed. These must be in the `include` directory. Defaults to all Stan
+#' files in the `include` directory. Note that the following files are
+#' overloaded and cannot be exposed: "delay_lpmf.stan",
+#' "allocate_observed_obs.stan", "obs_lpmf.stan", and "effects_priors_lp.stan".
+#'
+#' @param include A character string specifying the directory containing Stan
+#' files. Defaults to the 'stan/functions' directory of the [epinowcast()]
+#' package.
+#'
+#' @param global A logical value indicating whether to expose the functions
+#' globally. Defaults to `TRUE`. Passed to the [expose_functions()] method of
+#' [cmdstanr::CmdStanModel].
+#'
+#' @param ... Additional arguments passed to [enw_model]().
+#'
+#' @inheritParams enw_model
+#' @return An object of class `CmdStanModel` with functions from the model
+#' exposed for use in R.
+#'
+#' @family modeltools
+#' @importFrom cmdstanr write_stan_file
+#' @importFrom cli cli_abort cli_warn
+#' @export
+#' @examplesIf interactive()
+#' # Compile functions in stan/functions/hazard.stan
+#' stan_functions <- enw_stan_to_r("hazard.stan")
+#' # These functions can now be used in R
+#' stan_functions$functions$prob_to_hazard(c(0.5, 0.1, 0.1))
+#' # or exposed globally and used directly
+#' prob_to_hazard(c(0.5, 0.1, 0.1))
+enw_stan_to_r <- function(
+  files = list.files(include),
+  include = system.file("stan", "functions", package = "epinowcast"),
+  global = TRUE,
+  verbose = TRUE,
+  ...
+) {
+  overloaded_fns <- c(
```

Comment:
It's a shame to lose these. You *could* identify them with a tag (in comments) or split by new lines and then identify function names. But perhaps not work the effort / overhead.

Reply:
So identifying them in the comments won't work I think but I think having them have their own names (i.e. blah_blah_vect) would work and may be the way to go but I think perhaps its own issue to resolve?

Reply:
Sorry I wasn't very clear. You could identify overloaded functions to discard in the comments (or even have an `// *** R ignore from here` or the like which would be easy to filter out when reading in the functions). As for R e.g. you don't need `obs_lpmf(int ...)` as a length 1 vector can be passed to the array version. With this you would then still get a version of all the functions.